Alpha crash with shortest path query with password predicate

If you suspect this could be a bug, follow the template.

- What version of Dgraph are you using?

docker latest and play.dgraph.io

- Have you tried reproducing the issue with latest release?

Yes

- What is the hardware spec (RAM, OS)?

Various N/A

- Steps to reproduce the issue (command/config used to run Dgraph).

Send a query with a predicate containing at least one predicate of type password.
For instance, you can use the internal `dgraph.password`, but any other will also cause the issue.

```
{
  path as shortest(from: 0x1, to: 0x2) {
	  <dgraph.password>
  }
  path(func: uid(path)) {
  	name
	}
}
```

- Expected behaviour and actual result.

I expected to get a result, instead I get a panic:

```
panic: runtime error: index out of range

goroutine 2248 [running]:
github.com/dgraph-io/dgraph/query.(*SubGraph).expandOut(0xc0015d2000, 0x4f997e0, 0xc012fc8f30, 0xc0148591a0, 0xc012fc22a0, 0xc016d28a80)
	/Users/francesc/src/github.com/dgraph-io/dgraph/query/shortest.go:186 +0xfca
created by github.com/dgraph-io/dgraph/query.shortestPath
	/Users/francesc/src/github.com/dgraph-io/dgraph/query/shortest.go:455 +0x226
```

--- a/query/query3_test.go
+++ b/query/query3_test.go
@@ -482,7 +482,6 @@ func TestTwoShortestPathMinWeight(t *testing.T) {
 }
 
 func TestShortestPath(t *testing.T) {
-
 	query := `
 		{
 			A as shortest(from:0x01, to:31) {
@@ -515,6 +514,25 @@ func TestShortestPathRev(t *testing.T) {
 	require.JSONEq(t,
 		`{"data": {"_path_":[{"uid":"0x17","_weight_":1, "friend":{"uid":"0x1"}}],"me":[{"name":"Rick Grimes"},{"name":"Michonne"}]}}`,
 		js)
+}
+
+// Regression test for https://github.com/dgraph-io/dgraph/issues/3657.
+func TestShortestPathPassword(t *testing.T) {
+	query := `
+		{
+			A as shortest(from:0x01, to:31) {
+				password
+				friend
+			}
+
+			me(func: uid( A)) {
+				name
+			}
+		}`
+	js := processQueryNoErr(t, query)
+	require.JSONEq(t,
+		`{"data": {"_path_":[{"uid":"0x1", "_weight_": 1, "friend":{"uid":"0x1f"}}],
+			"me":[{"name":"Michonne"},{"name":"Andrea"}]}}`, js)
 }
 
 func TestFacetVarRetrieval(t *testing.T) {

--- a/query/shortest.go
+++ b/query/shortest.go
@@ -185,7 +185,7 @@ func (sg *SubGraph) expandOut(ctx context.Context,
 				for mIdx, fromUID := range subgraph.SrcUIDs.Uids {
 					// This can happen when trying to go traverse a predicate of type password
 					// for example.
-					if len(subgraph.uidMatrix) < mIdx+1 {
+					if mIdx >= len(subgraph.uidMatrix) {
 						continue
 					}
 

--- a/query/shortest.go
+++ b/query/shortest.go
@@ -183,6 +183,12 @@ func (sg *SubGraph) expandOut(ctx context.Context,
 
 				// Send the destuids in res chan.
 				for mIdx, fromUID := range subgraph.SrcUIDs.Uids {
+					// This can happen when trying to go traverse a predicate of type password
+					// for example.
+					if len(subgraph.uidMatrix) < mIdx+1 {
+						continue
+					}
+
 					for lIdx, toUID := range subgraph.uidMatrix[mIdx].Uids {
 						if adjacencyMap[fromUID] == nil {
 							adjacencyMap[fromUID] = make(map[uint64]mapItem)

